impl/sql/mutationstorage/queue.go

```diff
@@ -101,57 +102,70 @@ func (r *Receiver) run(ctx context.Context, last time.Time) {
 	}
 
 	if time.Since(last) > (r.opts.MaxPeriod - r.opts.Period) {
-		r.sendBatch(ctx, 0, r.opts.MaxBatchSize) // We will be overdue for an epoch soon.
+		if _, err := r.sendMultiBatch(ctx, 0, int(r.opts.MaxBatchSize)); err != nil {
+			glog.Errorf("firstTick: sendMultiBatch(): %v", err)
+		}
 	}
 
 	for {
-		var count int32
 		select {
-		case <-r.more:
-			count = r.sendBatch(ctx, 1, r.opts.MaxBatchSize)
 		case <-r.ticker.C:
-			count = r.sendBatch(ctx, 1, r.opts.MaxBatchSize)
+			if _, err := r.sendMultiBatch(ctx, 1, int(r.opts.MaxBatchSize)); err != nil {
+				glog.Errorf("minTick: sendMultiBatch(): %v", err)
+			}
 		case <-r.maxTicker.C:
-			count = r.sendBatch(ctx, 0, r.opts.MaxBatchSize)
+			if _, err := r.sendMultiBatch(ctx, 0, int(r.opts.MaxBatchSize)); err != nil {
+				glog.Errorf("maxTick: sendMultiBatch(): %v", err)
+			}
 		case <-ctx.Done():
 			return
 		case <-r.done:
 			return
 		}
-		if count >= r.opts.MaxBatchSize {
-			// Continue sending until we drop below batch size.
-			r.more <- time.Now()
+	}
+}
+
+// sendMultiBatch will send multiple batches if the number of available messages > maxMsgs.
+// This helps the queue catch up when there is high traffic, rather than waiting for the next
+// mintick to occur.
+func (r *Receiver) sendMultiBatch(ctx context.Context, minMsgs, maxMsgs int) (int, error) {
+	var total int
+	var err error
+	for sent := maxMsgs; sent >= maxMsgs; {
```

Comment:
If someone dumps a very large number of messages then it looks like this doesn't exit until they're all done. You might want to bound the number of batches per run.

Reply:
I *think* I actually want this to run as long as there a lot of messages to process. The rest of the code in this file implements waiting semantics.

Reply:
If there's 10,000,000 though? You could swamp the rest of the system. Though this might not be the right place to do that coordination.